### ISSUE TITLE
Library not loaded: @rpath/libopencv_core.3.4.dylib

### ISSUE THREAD

**Issue by fakob** [electron-userland/electron-builder#2738]
* **Version**: 20.2.1

* **Target**: mac

Hi, 
For [my project](https://github.com/fakob/MoviePrint_v004), which uses the [electron-react-boilerplate](https://github.com/chentsulin/electron-react-boilerplate) and electron-builder I am trying to create an executable which includes the opencv library. For this I am including the [opencv4nodejs](https://github.com/justadudewhohacks/opencv4nodejs) library which also builds and includes opencv.

Building, packaging and running it on the build machine works fine, but as soon as I try the executable on a different machine, I get the following error when executing the app.

```
Uncaught Exception:
Error: dlopen(/Applications/MoviePrint_v004.app/Contents/Resources/app.asar.unpacked/node_modules/opencv4nodejs/build/Release/opencv4nodejs.node, 1): Library not loaded: @rpath/libopencv_core.3.4.dylib
  Referenced from: /Applications/MoviePrint_v004.app/Contents/Resources/app.asar.unpacked/node_modules/opencv4nodejs/build/Release/opencv4nodejs.node
  Reason: image not found
    at process.module.(anonymous function) [as dlopen] (ELECTRON_ASAR.js:172:20)
    at Object.Module._extensions..node (module.js:598:18)
    at Object.module.(anonymous function) [as .node] (ELECTRON_ASAR.js:186:18)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Applications/MoviePrint_v004.app/Contents/Resources/app.asar/node_modules/opencv4nodejs/lib/opencv4nodejs.js:13:8)
    at Object.<anonymous> (/Applications/MoviePrint_v004.app/Contents/Resources/app.asar/node_modules/opencv4nodejs/lib/opencv4nodejs.js:31:3)
```

I have tried it within an asar file, then with `asarUnpack` for only `opencv-build` then also for `opencv4nodejs`, but with a similar result. 

Why do the last 2 paths in the error refer to the app.asar file, but the first 2 paths refer to the app.asar.unpacked folder?
How does using `asarUnpack` affect, if at all, how to link to libraries of another node_module? 

When extracting the app.asar file it seems that I get exactly the same structure. Is the app.asar file using symlinks and if so, could that be the problem?

```
Resources
--app.asar
    |--node_modules
        |--opencv-build
            |--opencv
                |--build
                    |--lib
                        |--opencv dylib files
        |--opencv4nodejs
            |--build
                |--Release
                    |--opencv4nodejs.node
--app.asar.unpacked
    |--node_modules
        |--opencv-build
            |--opencv
                |--build
                    |--lib
                        |--opencv dylib files
        |--opencv4nodejs
            |--build
                |--Release
                    |--opencv4nodejs.node
```

I am not fully sure how exactly the linking works, but I assumed that @rpath does not get resolved properly. 
So I tried to add different `rpath` paths, but I seem to not get it right.

**Comment by fakob** [comment]
Finally I made it work. 
What I had overlooked was the possibility to prevent packing the app folder into an asar file completely. I only had used the `asarUnpack` feature which seems to not include the symlinks. Now I have added this to the electron-builder setup
```
"asar": false
```
This combined with previously adding the proper relative path in `binding.gyp` for `opencv4nodejs.node` did the trick. 
```
"link_settings": {
	"libraries": [
		"-Wl,-rpath,@loader_path/../../../opencv-build/opencv/build/lib"
	],
},
```

**Comment by stoefln** [comment]
@fakob I am experiencing something very similar. Do you think this is related?
https://stackoverflow.com/questions/55593302/js-error-when-executing-electron-app-with-native-module-on-other-computer

**Comment by fakob** [comment]
Hi @stoefln 
Good question, I am not really sure. Especially because I did a lot of trial and error to figure it out. But this error
>Error: dlopen(/var/folders/7m/klqbsvis2iqg_mb_pdvv79fhO000gn/T/.app.repeato.uuz70R
looks familiar. If I remember correctly this seems to be the folder the app.asar temporarily unpacks the native package.

Have you tried setting `"asar": false`
This is [my project](https://github.com/fakob/MoviePrint_v004) where I made it work.

**Comment by stoefln** [comment]
thanks @fakob for the quick response. 
If I disable asar, I get 
```
/Users/steph/Documents/workspace/repeato3/node_modules/opencv4nodejs/build/Release/opencv4nodejs.node: Error: dlopen(/Users/steph/Documents/workspace/repeato3/node_modules/opencv4nodejs/build/Release/opencv4nodejs.node, 1): image not found
    at Object.<anonymous> (main-bundle.js:75)```

The path is correct, but it points to the original development folder of my project (renamed my dev folder, to see whether it would run on another computer...). I checked the line causing the error and it it references an absolute path indeed. I really wonder why...

**Comment by fakob** [comment]
Hi @stoefln  (saw that you are also a 'Wiener' :-)

I really tried hard to recollect what else I did or how you could solve this error. To be honest I never fully understood the linking so I am afraid that I am at the end of my wits. 

When researching this problem I also posted on [node-gyp](https://github.com/nodejs/node-gyp/issues/1397#issue-308194306). Have you tried running `otool -l opencv4nodejs.node`? Maybe this gives you a clue.

What I also do in my app is I [package ffmpeg libraries into the build](https://github.com/fakob/MoviePrint_v004/blob/master/scripts/ffmpeg.js) and relink them so they get recognised by opencv. This is probably not directly related to your error, but I still wanted to mention it. You might run into this issue as well in case you need opencv with ffmpeg support.

**Comment by stoefln** [comment]
@fakob thank you very much, will try to investigate further...

**Comment by champ8644** [comment]
Hi, @fakob 
I also building project using  electron-react-boilerplate and electron-builder with opencv4nodejs library and has kind of different problem. for my problem, in dev environment is going smoothly without any errors but if I package it into app and try running on the same computer - there is an error which I believe is related to something like opencv4nodejs is not loaded into the packed app.
this is snapshot of my error:
![image](https://user-images.githubusercontent.com/37543482/76689344-89ee9380-6667-11ea-9390-af308d143cc3.png)
Do you encounter something similar to this? and if so, how you manage with them? thank you very much.

**Comment by fakob** [comment]
Hi @champ8644 
Have you tried to add this to the electron-builder setup
```
"asar": false
```
The `appdata\Local\Temp\` path in your error is, I believe, used to temporarily unpack the asar package. Even though it is recommended to use asar, I could not make it work for opencv4nodejs as the library linking would break and therefore had to set it to false.

**Comment by stoefln** [comment]
@fakob AFAIR the node module can not be loaded because a dll dependency is missing. I think you helped me to solve this by copying `/Windows/system32/CONCRT140.dll` to the project directory during packaging. Interestingly I just saw that you removed this copy instructions from your package.json by now. I still have it in my project...

**Comment by fakob** [comment]
@champ8644 did you get it to work with my or @stoefln suggestion?

@stoefln, thanks for the tip. FYI, not sure when and why, but I had moved the copying of that dll into this script - https://github.com/fakob/MoviePrint_v004/blob/master/scripts/opencv.js

**Comment by stoefln** [comment]
@fakob - nice- good to know!

**Comment by champ8644** [comment]
@fakob Sorry for a very late reply, just `"asar": false` has the jobs done without the need to copy `/Windows/system32/CONCRT140.dll` to the project. Actually if you just copy the file without setting `"asar": false` it'll not work. 

But configure asar to false is adding another whopping 300 MB to my project - which I didn't know this is normal or not, but it's working so this is okay. Maybe this is just the whole opencv4nodejs builds is packed with my project.

**Comment by stoefln** [comment]
@champ8644 Regards bundle size: check out the copy files section in 
https://github.com/fakob/MoviePrint_v004/blob/master/package.json
@fakob figured out a lot of stuff which shouldn't go into the bundle. That brings down the size quite a bit.

**Comment by champ8644** [comment]
@stoefln @fakob wow nice! This really works - those 300 MB is gone and every other things is running as usual, That's a lot! Thanks you very much guys. Because previously I have tried to use electron-updater to update my app remotely but it takes forever to downloaded it. This would help me a lot. Btw I have tried differential update of electron-updater but it's no use so I can only update it by downloading the new installer everytime.
